File: julia_solutions/day04/day04.jl
module Day04

export run

# Function for the problem
function run(input::String)
    card_data = readinput(input)
    part_1 = part1(card_data)
    part_2 = part2(card_data)
    return [part_1, part_2]
end

# Function for part1 of the problem
function part1(card_data::Vector)
    points = 0
    for card in card_data
        count = [1 for number in card[2] if number in card[1]] |> sum
        points = count != 0 ? points + 2^(count-1) : points
    end
    return round(Int, points)
end

# Function for part2 of the problem
function part2(card_data::Vector)
    number_of_cards = ones(length(card_data))
    for (i, card) in enumerate(card_data)
        for _ in 1:number_of_cards[i]
            count = [1 for number in card[2] if number in card[1]] |> sum
            for j in i+1:i+count
                number_of_cards[j] += 1
            end
        end
    end
    return round(Int, sum(number_of_cards))
end

# Function for reading the input of the problem
function readinput(input::String)
    lines = readlines(input)
    card_data = []
    for line in lines
        numbers = split(split(line, ':')[2], '|')
        winning_numbers = parse.(Int, SubString.(numbers[1], findall(r"(\d+)", numbers[1])))
        player_numbers = parse.(Int, SubString.(numbers[2], findall(r"(\d+)", numbers[2])))
        push!(card_data, [winning_numbers, player_numbers])
    end
    return card_data
end

end

File: golang_solutions/day04/day04.go
package day04

import (
	"bufio"
	"log"
	"math"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

func Part1(filename string) int {
	cardData := readInput(filename)

	sum := 0
	for _, card := range cardData {
		count := 0
		for _, number := range card[1] {
			if slices.Contains(card[0], number) {
				count++
			}
		}
		if count > 0 {
			sum = sum + int(math.Pow(2, float64(count-1)))
		}
	}

	return sum
}

func Part2(filename string) int {
	cardData := readInput(filename)

	numberOfCards := make([]int, len(cardData))
	for i := range numberOfCards {
		numberOfCards[i] = 1
	}

	for index, card := range cardData {
		for i := 0; i < numberOfCards[index]; i++ {
			count := 0
			for _, number := range card[1] {
				if slices.Contains(card[0], number) {
					count++
				}
			}
			for j := index + 1; j < index+count+1; j++ {
				numberOfCards[j]++

			}
		}

	}

	sum := 0
	for _, n := range numberOfCards {
		sum = sum + n
	}

	return sum
}

func readInput(filename string) [][][]int {

	// Open input file
	readFile, err := os.Open(filename)

	if err != nil {
		log.Fatal(err)
	}

	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)

	cardData := make([][][]int, 0)
	numbersRegex := regexp.MustCompile(`(\d+)`)

	for fileScanner.Scan() {
		line := fileScanner.Text()
		splitedLine := strings.Split(line, ":")
		splitedNumbers := strings.Split(splitedLine[1], "|")

		cardNumbers := make([][]int, 2)

		// Extracting winning numbers
		winningNumbersMatches := numbersRegex.FindAllStringSubmatch(splitedNumbers[0], -1)
		winningNumbers := make([]int, len(winningNumbersMatches))
		for i, match := range winningNumbersMatches {
			number, _ := strconv.Atoi(match[1])
			winningNumbers[i] = number
		}
		cardNumbers[0] = winningNumbers

		// Extracting player numbers
		playerNumbersMatches := numbersRegex.FindAllStringSubmatch(splitedNumbers[1], -1)
		playerNumbers := make([]int, len(playerNumbersMatches))
		for i, match := range playerNumbersMatches {
			number, _ := strconv.Atoi(match[1])
			playerNumbers[i] = number
		}
		cardNumbers[1] = playerNumbers

		cardData = append(cardData, cardNumbers)
	}

	return cardData

}
